### DISCUSSION TITLE
Modern hashes

### DISCUSSION THREAD

**Discussion by jpellegrini** [egallesio/STklos#187] [Ideas]
Hi @egallesio !

STklos has `md5sum`, and it would be nice to have more modern hash functions.

What about, instead of adding more to the STklos core, creating a "hash" module?
I already have one SHA3-512 implementation, just as a proof of concept. I gues it would be nice to have:
* SHA3 family, including SHAKE128 and SHAKE256
* SHA256 at least (because it's widely used), and maybe (what do you think?) the other SHA-2 variants?

What about the API?

```
(sha3-512 string)    ; one function per variant
(sha3 string 512)    ; one function per family, object first
(sha3 512 string)    ; one function per family, # of bits first
```

Which sounds better? Maybe have all of them?

I suppose it would not be good to remove MD5 from the core and make it a module?

**Comment by jpellegrini** [comment]
Also, what about having separate SHA-2 and SHA-3 packages? For space-constrained environments...

**Reply by egallesio** [reply]
> What about, instead of adding more to the STklos core, creating a "hash" module?

Good idea.

`(sha3 string 512)` is porbably preferable since, for instance, 512 can be a default value for the variant. Hence,   `(sha3 string)`  would be sufficient to use  SHA3-512. 

> I suppose it would not be good to remove MD5 from the core and make it a module?

Probably

**Comment by lassik** [comment]
HTTPS is increasingly required for web access, and OpenSSL/LibreSSL include most of the common cryptogrpahic hash functions. It would make sense to re-use those implementations, which are probably fast and battle-tested.

**Reply by jpellegrini** [reply]
Yes... But a smaller implementation would be fine -- the algorithm is not long, it can be checked, and there is a test suite from NIST. If it passes, I wouldn't expect it to be broken in any way. And having a standalone implementation (without OpenSSL/etc) is nice, since it's smaller.

But an "openssl" module via ScmPkg, with everything that OpenSSL offers (I know, it's a huge task) would be nice also.

**Comment by lassik** [comment]
For example, [here are the SHA implementations](https://github.com/libressl-portable/openbsd/tree/master/src/lib/libcrypto/sha).
